<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <script>
        /**
         * 链式编程-处理异常和普通内容(fulfilled状态)
         * 1. 返回新的Promise实例
         * 2. 获取返回值
         *   2.1 处理返回值
         *   2.2 处理异常
        */
        function runAsynctask(callback) {
        // 2. 调用核心api(queueMicrotask,MutationObserver,setTimeOut)
        if(typeof queueMicrotask === 'function') {
            queueMicrotask(callback)
        }else if(MutationObserver === 'function') {
            const obs = new MutationObserver(callback)
            const divNode = document.createElement('div')
            obs.observe(divNode,{childList:true})
            divNode.innerHTML = 'kkkkk'
        }else{
            setTimeout(callback,0)
        }
       }
        const PENDING = 'pending'
        const FULFILLED = 'fulfilled'
        const REJECTED = 'rejected'
        class MyPromise {
            // 添加状态
            state = PENDING
            // 添加原因
            result = undefined

            //定义实例属性
            #handlers = [] // [{onFulfilled,onRejected}...]

            //构造函数
            constructor(fun) {
                // 调整resolve/reject
                const resolve = (result) => {
                    // 状态不可逆
                    if(this.state === PENDING) {
                        // 改状态：pending->fulfilled
                        this.state = FULFILLED
                        // 记录原因
                        this.result = result
                        // 3.调用成功的回调
                        this.#handlers.forEach(({onFulfilled}) => {
                            onFulfilled(this.result)
                        })
                    }
                }
                const reject = (result) => {
                    // 状态不可逆
                    if(this.state === PENDING) {
                        // 改状态：pending->rejected
                        this.state = REJECTED
                        // 记录原因
                        this.result = result
                        // 4. 调用失败的回调
                        this.#handlers.forEach(({onRejected}) => {
                            onRejected(this.result)
                        })
                    }
                }
                fun(resolve,reject)
            }
            // 添加实例方法
            then(onFulfilled,onRejected){
                // 参数判断
                onFulfilled = typeof onFulfilled === 'function'? onFulfilled : x => x
                onRejected = typeof onRejected === 'function'? onRejected : x => {throw x}

                // 1. 返回新的Promise实例
                const p2 = new MyPromise((resolve,reject) => {
                    if (this.state === FULFILLED){
                    runAsynctask(() => {
                        try {
                            //2. 获取返回值
                            const x =  onFulfilled(this.result)
                            // 2.1 处理返回值
                            resolve(x)
                        } catch (error) {
                            // 2.2 处理异常
                            reject(error)
                        }
                    })
                }else if(this.state === REJECTED){
                    runAsynctask(() => {
                        onRejected(this.result)
                    })
                }else if(this.state === PENDING){
                    // 保存回调函数
                    this.#handlers.push({
                        onFulfilled:()=>{
                            runAsynctask(() => {
                                onFulfilled(this.result)
                            })
                        },
                        onRejected:()=>{
                            runAsynctask(() => {
                                onRejected(this.result)
                            })
                        }
                    })
                }
                })
                return p2
            }
        }


        // -----------------测试代码----------------------
        const p = new MyPromise((resolve,reject) => {
            resolve(1)
        })
        p.then(res => {
            console.log('p1',res)
            throw 'throw-error'
            return 2
        }).then(res => {
            console.log('p2',res)
        },err => {
            console.log('p2',err)
        })
    </script>
</body>
</html>